core/app/models/spree/promotion.rb

```diff
@@ -26,31 +31,45 @@ class Promotion < Spree::Base
 
     before_save :normalize_blank_values
 
-    scope :coupons, ->{ where("#{table_name}.code IS NOT NULL") }
-
     order_join_table = reflect_on_association(:orders).join_table
-
-    scope :applied, -> { joins("INNER JOIN #{order_join_table} ON #{order_join_table}.promotion_id = #{table_name}.id").uniq }
+    scope :applied, -> do
+      joins("INNER JOIN #{order_join_table} ON " \
+        "#{order_join_table}.promotion_id = #{table_name}.id").distinct
+    end
+    scope :coupons, -> do
+      joins(:promotion_codes).where("#{Spree::PromotionCode.table_name}.value IS NOT NULL")
+    end
 
     def self.advertised
       where(advertise: true)
     end
 
     def self.with_coupon_code(coupon_code)
-      where("lower(#{self.table_name}.code) = ?", coupon_code.strip.downcase).first
+      joins(:promotion_codes).where(
+        "lower(#{Spree::PromotionCode.table_name}.value) = ?", coupon_code.strip.downcase
```

Comment:
You might also consider:

```
joins(:promotion_codes).where(
  Spree::PromotionCode.arel_table[:value].matches(coupon_code.strip)
)
```

Even better. You could normalize values going in to the table as part of the migration (such that the values stored in the DB are always lower case), and normalize them in the PromotionCode class. Then the DBMS can use the index for lookups.